ocaml/typing/outcometree.mli

```diff
@@ -108,13 +108,15 @@ type out_jkind_user =
   | Ojkind_user_mod of out_jkind_user * string list
   | Ojkind_user_with of out_jkind_user * out_type
   | Ojkind_user_kind_of of out_type
+  | Ojkind_user_arrow of out_jkind_user Jkind_types.Arrow.t
 
 and out_jkind_const = { base : string; modal_bounds : string list }
 
 and out_jkind =
   | Ojkind_user of out_jkind_user
   | Ojkind_const of out_jkind_const
   | Ojkind_var of string
+  | Ojkind_arrow of out_jkind Jkind_types.Arrow.t
```

Comment:
I think it would be better not to reuse `Arrow.t` here. These types represent the syntax of OCaml and the stuff in `Jkind` is the semantics. It is usually better to keep the types representing these things distinct, as they have different purposes and should generally contain different data. Using the same type tends to lead to either needing to separate them again later, or worse failing to separate them even once the semantic and syntactic data has diverged.

Reply:
I do agree, but I'm not sure I see why you consider `Jkind_types` to be part of the semantics? My intuition was that `Jkind_types` is syntax, and `Jkind` is for semantics. Thus `Types` (mostly syntax) and `Jkind` (semantics) depend on `Jkind_types` (syntax).

Reply:
>  `Types` (mostly syntax)

`Types` is the semantics of the type language. The syntax of e.g. type expressions lives in `Typedtree`. There is also a second copy of the syntax over in `Outcometree` used for printing.

`Jkind_types` is just the types used in `Jkind`, the split there is just to avoid a dependency cycle.

Reply:
Is it that you consider `Jkind` and `Jkind_types` notionally the same as they are just there to break the reference cycle?

Reply:
I do consider them as such, and their documentation comments say as much. But even were that not the case the types in `Jkind_types` are still describing the semantic kinds and not the syntax for kinds written by users (or printed in error messages).